torchvision/models/quantization/mobilenetv3.py

```diff
@@ -46,7 +46,11 @@ def _load_from_state_dict(
         if hasattr(self, "qconfig") and (version is None or version < 2):
             default_state_dict = {
                 "scale_activation.activation_post_process.scale": torch.tensor([1.0]),
+                "scale_activation.activation_post_process.activation_post_process.scale": torch.tensor([1.0]),
                 "scale_activation.activation_post_process.zero_point": torch.tensor([0], dtype=torch.int32),
+                "scale_activation.activation_post_process.activation_post_process.zero_point": torch.tensor(
+                    [0], dtype=torch.int32
+                ),
```

Comment:
@vkuzo @jerryzh168 FYI, we believe that after PR https://github.com/pytorch/pytorch/pull/68143 this is now necessary to make the weights compatible.

The minimal code to verify the difference is comparing the output of:
```python
import torch
from torchvision.models.quantization.mobilenetv3 import _mobilenet_v3_conf, _replace_relu, QuantizableMobileNetV3, QuantizableInvertedResidual

inverted_residual_setting, last_channel = _mobilenet_v3_conf("mobilenet_v3_large")
m = QuantizableMobileNetV3(inverted_residual_setting, last_channel, block=QuantizableInvertedResidual)
_replace_relu(m)
m.fuse_model()
m.qconfig = torch.quantization.get_default_qat_qconfig("qnnpack")
torch.quantization.prepare_qat(m, inplace=True)

print("\n".join(m.state_dict().keys()))
```

The records in question don't exist on earlier versions of pytorch nightly. Just wanted to make sure that this is an expected behaviour.

Reply:
It seems that the situation is much more broken than we thought. The quantization scripts of TorchVision no longer work using the latest nightly of Core (`pytorch                   1.11.0.dev20211125 py3.9_cuda11.1_cudnn8.0.5_0    pytorch-nightly`):
```
$ python train_quantization.py --device='cpu' --post-training-quantize --backend='fbgemm' --model='shufflenet_v2_x0_5'
Not using distributed mode
Namespace(data_path='/datasets01/imagenet_full_size/061417/', model='shufflenet_v2_x0_5', backend='fbgemm', device='cpu', batch_size=32, eval_batch_size=128, epochs=90, num_observer_update_epochs=4, num_batch_norm_update_epochs=3, num_calibration_batches=32, workers=16, lr=0.0001, momentum=0.9, weight_decay=0.0001, lr_step_size=30, lr_gamma=0.1, print_freq=10, output_dir='.', resume='', start_epoch=0, cache_dataset=False, sync_bn=False, test_only=False, post_training_quantize=True, world_size=1, dist_url='env://', interpolation='bilinear', val_resize_size=256, val_crop_size=224, train_crop_size=224, clip_grad_norm=None, weights=None, distributed=False)
Loading data
Loading data
Loading training data
Took 5.208142042160034
Loading validation data
Creating data loaders
Creating model shufflenet_v2_x0_5
./torch/ao/quantization/observer.py:177: UserWarning: Please use quant_min and quant_max to specify the range for observers.                     reduce_range will be deprecated in a future release of PyTorch.
  warnings.warn(
Calibrating
Traceback (most recent call last):
  File "./vision/references/classification/train_quantization.py", line 274, in <module>
    main(args)
  File "./vision/references/classification/train_quantization.py", line 103, in main
    evaluate(model, criterion, data_loader_calibration, device=device, print_freq=1)
  File "./vision/references/classification/train.py", line 76, in evaluate
    output = model(image)
  File "./torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "./vision/torchvision/models/quantization/shufflenetv2.py", line 49, in forward
    x = self._forward_impl(x)
  File "./vision/torchvision/models/shufflenetv2.py", line 145, in _forward_impl
    x = self.conv1(x)
  File "./torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "./torch/nn/modules/container.py", line 141, in forward
    input = module(input)
  File "./torch/nn/modules/module.py", line 1131, in _call_impl
    hook_result = hook(self, input, result)
  File "./torch/ao/quantization/quantize.py", line 91, in _observer_forward_hook
    return self.activation_post_process(output)
  File "./torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "./torch/ao/quantization/observer.py", line 1093, in forward
    combined_histogram = torch.histc(
RuntimeError: Inplace update to inference tensor outside InferenceMode is not allowed.You can make a clone to get a normal tensor before doing inplace update.See https://github.com/pytorch/rfcs/pull/17 for more details.
```

I have confirmed that there are no inplace operations on the network (all of them are replaced prior quantizing).



Installing `pytorch                   1.11.0.dev20211020 py3.9_cuda11.1_cudnn8.0.5_0    pytorch-nightly` works. 

I also tried recent nightlies from the last week or so, but doesn't work. Something broke it earlier this month. 

cc @raghuramank100

Reply:
> @vkuzo @jerryzh168 FYI, we believe that after PR pytorch/pytorch#68143 this is now necessary to make the weights compatible.

cc @andrewor14, @jerryzh168 , can we adjust the logic from https://github.com/pytorch/pytorch/pull/68143 to enable loading weights from the previous format?  It should be possible with state dict versioning.

Reply:
> It seems that the situation is much more broken than we thought. The quantization scripts of TorchVision no longer work using the latest nightly of Core (pytorch 1.11.0.dev20211125 py3.9_cuda11.1_cudnn8.0.5_0 pytorch-nightly):

@z-a-f  who is oncall this week, could you take a look or help find someone to take a look?

Reply:
@vkuzo, @andrewor14, @jerryzh168, @z-a-f Please let me know what's the issue that tracks this problem and if there is a workaround we can apply temporarily to fix the broken code on TorchVision.

Reply:
Hi @datumbox. I will look into this issue and track our progress using the issue you opened (https://github.com/pytorch/vision/issues/5005)

Reply:
> 

Did you try bisecting the pytorch to see the offender? If now, I guess I can quickly run through it

Reply:
I created an issue with Pytorch: https://github.com/pytorch/pytorch/issues/69235, we'll track it there

>

Reply:
I was able to reproduce the above issues on my end. Just to be clear, there are two different issues here:

1) Extra entries in state_dict that this PR addresses. This was directly caused by 79b67d9. I have opened https://github.com/pytorch/pytorch/issues/69246 to track this.

2) Quantization scripts in torchvision throws "inplace update not allowed" error. This was not related to 79b67d9 because the it's also present without that commit. This is tracked by https://github.com/pytorch/pytorch/issues/69235 and https://github.com/pytorch/vision/issues/5005.

Reply:
I think the first one is expected, if it's introduced by replacing FixedQParamFakeQuantize with FakeQuantize + FixedQParamObserver, although it looks like this is eager mode quantization and I'm not sure why we are using it here, I remember we only have it in fx graph mode quantization.

Reply:
> Extra entries in state_dict that this PR addresses. This was directly caused by 79b67d9. I have opened this issue to track this.

@andrewor14 Are you treating this as a bug? If yes, then I assume we will need to revert this PR.

Reply:
Hi @datumbox, after an offline discussion with @jerryzh168 we decided the `state_dict` issue is not really a bug so we probably won't fix it for now. In the future if we decide to remove `FixedQParamsFakeQuantize` as part of a cleanup then we may have to revert this PR.

Reply:
@andrewor14 Thanks for the clarification. Happy to leave the patch of this PR in place until you indicate otherwise.

Could you please confirm that you are still looking for resolving the "inplace update not allowed" error?

Reply:
@datumbox I'll take a look now

Reply:
Hi @datumbox I can't repro the error with torchvision and pytorch master , could you share a repro PR?

Reply:
@jerryzh168 Actually I just tried it with master (bcd0303) and I was able to reproduce it using the command in https://github.com/pytorch/vision/pull/4997#discussion_r757569785

Reply:
That's right. Just run:
```
cd references/classification
python train_quantization.py --device='cpu' --post-training-quantize --backend='fbgemm' --model='shufflenet_v2_x0_5'
``` 

Let me know if you can reproduce it.

Reply:
trying to repro with master, but currently I have this error:
```
Traceback (most recent call last):
  File "train_quantization.py", line 12, in <module>
    from train import train_one_epoch, evaluate, load_data
  File "/data/users/jerryzh/vision/references/classification/train.py", line 12, in <module>
    from references.classification.sampler import RASampler
ModuleNotFoundError: No module named 'references'
```
it's introduced in some recent PRs I think, could you help fix it?

Reply:
Ah yes, it was reported earlier but the fix wasn't submitted yet. I just merged it, could you try again?

Reply:
thanks, I just pulled the most recent master and it works now. I can't repro the issue. Could you try again?

Reply:
I can run 
```
python train_quantization.py --device='cpu' --post-training-quantize --backend='fbgemm' --model='shufflenet_v2_x0_5' --data-path=$DATA_PATH
```

```
...
Test:   [29/32]  eta: 0:00:00  loss: 1.1107 (1.0689)  acc1: 78.1250 (79.5833)  acc5: 87.5000 (88.4375)  time: 0.1065  data: 0.0001  max mem: 0
Test:   [30/32]  eta: 0:00:00  loss: 1.1233 (1.0709)  acc1: 78.1250 (79.6371)  acc5: 87.5000 (88.4073)  time: 0.1064  data: 0.0001  max mem: 0
Test:   [31/32]  eta: 0:00:00  loss: 1.1233 (1.0694)  acc1: 78.1250 (79.7852)  acc5: 87.5000 (88.4766)  time: 0.1057  data: 0.0001  max mem: 0
Test:  Total time: 0:00:07
Test:  Acc@1 79.785 Acc@5 88.477
/home/jerryzh/anaconda3/envs/py3/lib/python3.9/site-packages/torch/ao/quantization/observer.py:899: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  src_bin_begin // dst_bin_width, 0, self.dst_nbins - 1
/home/jerryzh/anaconda3/envs/py3/lib/python3.9/site-packages/torch/ao/quantization/observer.py:904: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  src_bin_end // dst_bin_width, 0, self.dst_nbins - 1
Saving quantized model
Evaluating post-training quantized model
Test:   [  0/391]  eta: 0:18:59  loss: 0.8914 (0.8914)  acc1: 79.6875 (79.6875)  acc5: 90.6250 (90.6250)  time: 2.9136  data: 2.4820  max mem: 0
Test:   [100/391]  eta: 0:01:18  loss: 1.5384 (1.4906)  acc1: 59.3750 (64.7509)  acc5: 88.2812 (86.2005)  time: 0.2493  data: 0.0002  max mem: 0
Test:   [200/391]  eta: 0:00:55  loss: 2.5032 (1.6645)  acc1: 47.6562 (62.2785)  acc5: 69.5312 (83.7881)  time: 0.4037  data: 0.0003  max mem: 0
Test:   [300/391]  eta: 0:00:26  loss: 2.5260 (1.8708)  acc1: 52.3438 (59.0713)  acc5: 70.3125 (80.5804)  time: 0.2374  data: 0.0002  max mem: 0
Test:  Total time: 0:01:47
Test:  Acc@1 58.124 Acc@5 79.728
```

Reply:
@jerryzh168 I confirm that something fixed it on the latest pytorch core. This is very strange. We can close the ticket and keep an eye on it in case it pops up again. Thanks for the help!